Elasticsearch: Adds support for region annotations

fangel opened the pull request:
It is now possible to specify a field containing a region-id in Elasticsearch annotations.
Two annotations events sharing the same region-id will be displayed as a single region
annotation. 
In order to support regions that start or end outside of the displayed time
frame, the region-id field will have to contain the document-id of the start-event

Ref #10589 for Elasticsearch based annotations

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=17602) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=17602) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=17602) it.</sub>

fangel commented:
A quick screenshot showing region that begin before the time range, a region fully contained within the time range, a non-region event, and a region that ends outside the time-range:

<img width="809" alt="Screenshot 2019-06-16 at 12 04 53" src="https://user-images.githubusercontent.com/19275/59562620-1fa0b800-902f-11e9-90ef-12685c73f5fd.png">

fangel commented:
This issue doesn't fully fix #10589, as it's only a single DS (Elasticsearch) that is introduced in this PR. But I don't know how to best write that in the commit/issue-msg.

marefr commented:
Thanks. This looks interesting. Will review it as soon as I get some time. You can write `Ref #10589` instead of `Fixes #10589`.

ryantxu commented:
rather than adding the akward regionId structure to datasources, maybe we can just add an optional endTime on all annotations.  see #17647

marefr commented:
> rather than adding the akward regionId structure to datasources, maybe we can just add an optional endTime on all annotations

If we change the api/model for annotations I agree that using an endTime is preferred, but currently the model only supports `regionId` and changing that is not something this PR should address.

ryantxu commented:
> If we change the api/model...

Exactly -- I think we should change the api/model rather than adding more ways to work with them as is.

See #17673 for my first pass at changing the backend

My bigger goal is that any query to any datasource that returns the fields we want (time, endTime?, title, etc) should be able to be used as annotations -- rather than requiring each datasource to implement a special way to support them.  See also #17280

fangel commented:
I agree that the changes to drop `regionId` and move to all annotations being able to have a `timeEnd` seems to simplify a lot of code-paths (based on a quick skim of the changes in #17673).

The changes to the ES datasource for mapping a `timeEnd` field instead of a `regionId`-field should be very minimal, so once the changes away from `regionId` lands, I can update this PR to target `timeEnd` instead.

marefr commented:
@fangel #17673 have been merged so I guess you should be able to use the new time and endTime props, right @ryantxu ?

fangel commented:
Sounds great. I'll try and re-create this PR either Thursday or over the weekend.

fangel committed:
Elasticsearch: Adds support for region annotations

It is now possible to specify a field containing time-end in Elasticsearch annotations.
Any annotations with a time-end will become a region between time and time-end. Any
annotations without the time-end field will remain a single-point annotation.

Ref #10589 for Elasticsearch based annotations

fangel commented:
Okay, here's the first attempt at using the new region method.

It took a bit more than the previous, because it now needs to also query for annotations with a time-end timestamp inside the current range.

As I noted in #17673, it also appears that the updates there actually breaks ES annotations (without the changes in the PR), because all ES annotations only sets the `time`, and not the `timeEnd`-property on the events.
These changes works around that, by always ensuring that a `timeEnd`-property is set - if no time-end field is specified, or the field is empty, then it simply reuses the value of the time-field.

fangel commented:
I rebase onto the latest master now, and made it so that timeEnd is only sent if there is a timeEnd.

However, that meant I had to help the type-checked a bit, because TS doesn't like `foo = {...}; foo.newProp = ...`.
It would probably be a beneficial if there was a shared definition of a annotation-type somewhere, but that seems beyond the scope of this PR.

marefr commented on public/app/plugins/datasource/elasticsearch/datasource.ts:
```diff
@@ -130,22 +131,39 @@ export class ElasticDatasource extends DataSourceApi<ElasticsearchQuery, Elastic
 
     const data: any = {
       query,
+      sort: [sort],
```
Why sort ascending, isn't that the default?

fangel replied:
First the reason why I added the sort-info in general:

I believe the default is actually to sort on score, then date - and because the old way used a non-scoring `filter`-query, the scores were always 0, and then the sort ended up being the date-field.

In order to search for annotations where either the start or end-time of the annotation-region are inside the current dasbhard-range, then the query had to be modified to a `should` bool-query with a `min_should_match: 1`-param.

Because the `should`-filter is a scoring filter, then as I understand it, annotations with both the start and end-data inside the region would score higher, and thus come out before regions with one the start or end-date inside the range.

---

And then you're actual question:

You are correct in that the default sorting is `asc`, so the code could probably be simplified into simply `sort: [ timeField ]` instead of `sort: [ {[timeField]: 'asc' }]` but I personally like the explicitness of having the sort-order in the code. 

But if you would prefer the simpler code relying on the default-sort, I'll swap it out.

marefr replied:
Thanks. Wouldn't be more correct to use two range-filters, one for `from` of `timeField` and one for `to` of `timeFieldEnd` if both `timeField` and `timeFieldEnd` are configured?

fangel replied:
Hmm, that might work, yes - I'll have to test that out. I'll probably have time to test it tomorrow.

If you a `timeField` and `timeFieldEnd` is defined, it currently builds a query like so:

```json
{
  "bool": {
    "should": [
      { "range": { "@timestamp": { "from": 0, "to": 2, "format": "epoch_millis" } } },
      { "range": { "@timestampEnd": { "from": 0, "to": 2, "format": "epoch_millis" } } }
    ],
    "minimum_should_match": 1
  }
}
```

But then it turns into a scoring query (because they are in the `should`-part of the `bool`-filter) - I hadn't considered nesting two `bool`-queries to make the `should`-query into a non-scoring query.

I don't know which query is the cleanest looking - the un-nested one with a sort, or a nested one that doesn't need a sort.
But there's a chance the nested one is faster, because it doesn't need to perform the sort, so if that works, I could be tempted into leaning towards the nested option you propose.

marefr reviewed (commented):
Thanks. Looks good :+1: Left a question that I would like you to ask and then we'll need to do a manual test of this.

fangel commented:
I got it work with a nested query, so the full query it builds is inside filter, and that the sort-clause have been removed.

A full example of the query that gets sent to ES is now:

```json
{
  "query": {
    "bool": {
      "filter": [
        { "query_string": {
            "query": "[QS]"
        } },
        { "bool": {
            "should": [
              { "range": {
                "[timeField]": {
                "from": [range-start],  "to": [range-end],  "format": "epoch_millis"
              }
              } },
              { "range": {
                "[timeFieldEnd]": {
                  "from": [range-start], "to": [range-end], "format": "epoch_millis"
                }
              } }
            ],
            "minimum_should_match": 1
        } }
      ]
    }
  },
  "size": 10000
}
```

So it's now a daily small change in the sent query (the `range`-filter is now a `bool.should`-filter with two `range`-clauses inside of it)

What do you think?

marefr reviewed (approved):
Sorry it taken me long time to come back to this one. LGTM

marefr commented:
Thank you for contributing to Grafana!

fangel commented:
You all are welcome! Thanks for a great product.
